bindings/rust/s2n-tls-tokio/src/lib.rs

```diff
@@ -190,21 +198,23 @@ where
         // We protect against this by pinning the stream during the action
         // and clearing the context afterwards.
         unsafe {
-            let context = self as *mut Self as *mut c_void;
+            let mut context = RxTxContext {
+                stream: Pin::new(&mut self.stream),
+                async_context: ctx
+            };
+            let context = (&mut context as *mut RxTxContext<S>).cast();
 
             self.as_mut().set_receive_callback(Some(Self::recv_io_cb))?;
             self.as_mut().set_send_callback(Some(Self::send_io_cb))?;
             self.as_mut().set_receive_context(context)?;
             self.as_mut().set_send_context(context)?;
-            self.as_mut().set_waker(Some(ctx.waker()))?;
```

Comment:
This change won't work with the async callbacks, who all expect there to be a waker set on the connection to work properly. Instead of changing the `context`, it should be possible to implement a stack `Waker` that holds a pinned pointer to the async context on the stack.

If this isn't possible, we need to change all of the functions to require `&mut core::task::Context` be passed into them. And this wouldn't be the worst thing; it would definitely make it more rust-like. It just breaks the signature.